src/value.rs

```diff
@@ -0,0 +1,869 @@
+                let values = de::impls::VecVisitor::new().visit_seq(visitor)?;
+                Ok(Value::Array(values))
+            }
+
+            fn visit_map<V>(self, mut visitor: V) -> Result<Value, V::Error>
+                where V: de::MapVisitor
+            {
+                let mut key = String::new();
+                let datetime = visitor.visit_key_seed(DatetimeOrTable {
+                    key: &mut key,
+                })?;
+                match datetime {
+                    Some(true) => {
+                        let date: DatetimeFromString = visitor.visit_value()?;
+                        return Ok(Value::Datetime(date.value))
+                    }
+                    None => return Ok(Value::Table(BTreeMap::new())),
+                    Some(false) => {}
+                }
+                let mut map = BTreeMap::new();
+                map.insert(key, visitor.visit_value()?);
+                while let Some(key) = visitor.visit_key()? {
+                    if map.contains_key(&key) {
+                        let msg = format!("duplicate key: `{}`", key);
+                        return Err(de::Error::custom(msg))
+                    }
+                    map.insert(key, visitor.visit_value()?);
+                }
+                Ok(Value::Table(map))
+            }
+        }
+
+        deserializer.deserialize(ValueVisitor)
+    }
+}
+
+impl de::Deserializer for Value {
+    type Error = ::de::Error;
+
+    fn deserialize<V>(self, visitor: V) -> Result<V::Value, ::de::Error>
+        where V: de::Visitor,
+    {
+        match self {
+            Value::Boolean(v) => visitor.visit_bool(v),
+            Value::Integer(n) => visitor.visit_i64(n),
+            Value::Float(n) => visitor.visit_f64(n),
+            Value::String(v) => visitor.visit_string(v),
+            Value::Datetime(v) => visitor.visit_string(v.to_string()),
+            Value::Array(v) => {
+                let len = v.len();
+                let mut deserializer = SeqDeserializer::new(v);
+                let seq = visitor.visit_seq(&mut deserializer)?;
+                let remaining = deserializer.iter.len();
+                if remaining == 0 {
+                    Ok(seq)
+                } else {
+                    Err(de::Error::invalid_length(len, &"fewer elements in array"))
+                }
+            }
+            Value::Table(v) => {
+                let len = v.len();
+                let mut deserializer = MapDeserializer::new(v);
+                let map = visitor.visit_map(&mut deserializer)?;
+                let remaining = deserializer.iter.len();
+                if remaining == 0 {
+                    Ok(map)
+                } else {
+                    Err(de::Error::invalid_length(len, &"fewer elements in map"))
+                }
+            }
+        }
+    }
+
+	fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, ::de::Error>
+		where V: de::Visitor
+	{
+		visitor.visit_some(self)
```

Comment:
Add a comment here and the other deserialize_option methods that this is because TOML does not have a representation corresponding to null.